context/src/main/java/io/opentelemetry/context/ContextStorage.java

```diff
@@ -81,4 +83,6 @@ static ContextStorage get() {
    * this will be the {@linkplain Context#root()} root context}.
    */
   Context current();
+
+  void onAttach(Consumer<Context> contextConsumer);
```

Comment:
I think I would take a approach that allows registering wrappers - as long as they are registered early i the app before storage is initialized, it allows customizing without thread sync issues.

```java

public interface ContextStorage {
  default void wrap(Function<ContextStorage, ContextStorage> wrapper) {
    ContextStorageWrappers.add(wrapper);
  }
}

LazyStorage {
  ContextStorage contextStorage = createStorage();
  for (Function wrapper : ContextStorageWrappers.get()) {
    contextStorage = wrapper.apply(contextStorage);
  }
}
```

What do you think?

Reply:
Registering decorators for the `ContextStorage` would be an interesting approach.  How would you get to the `ContextStorage` to call `wrap` without triggering `LazyStorage` to create the storage, though?

Reply:
`LazyStorage` will only be loaded when `get` is called so I think it should be ok :-D

Reply:
Got it, that makes sense now.  My only concern with this approach is that you only have that window of time during which you can register these decorators.  Perhaps `ContextStorage#wrap` should warn or throw if the storage has already been created?  Should there also be an SPI for ContextStorageWrappers?

Reply:
Maybe it's sufficient to just make it easy to wrap `ThreadLocalContextStorage` via SPI.

The Javadocs on `ContextStorage` imply that you can access the instance of `ThreadLocalContextStorage` via `Context#threadLocalStorage` but that method doesn't exist and I don't see another way to access that instance.  Without the ability to reference that instance rolling your own ContextStorage would involve rewriting thread-local storage which is where I think the opportunities for error mostly lie.

Reply:
I've opened #1972 as an alternative to this PR.  All it does is expose the `ThreadLocalContextStorage` instance to make it easier to roll your own basic `ContextStorage` implementation.

Reply:
Exposing ThreadLocalContextStorage was definitely required (it was javadoc'd, just the method got lost during some refactoring 😅 ).

I think `wrap` can help where SPI is too tedious, or to allow wrapping the non-threadlocal storage, but if our current code satisfies your use case well, we can consider that later. I think we could add it now though if it helps at all (it was the very first ask from users when we implemented storage SPI in Armeria so I do expect it to happen)

Reply:
The SPI approach scratches my itch for now, and now that ThreadLocalContextStorage is exposed again it makes my implementation that much simpler.   As for establishing an API here I think it worthwhile.  I'd be curious what the other use cases might be.

Reply:
Here's some reference of what led to the API in Armeria

https://github.com/line/armeria/issues/2634

In particular it seems, from my interpretation, to be about having a similar API for hooking as reactor for users of both.

Reply:
@HaloFour FYI sent out https://github.com/open-telemetry/opentelemetry-java/pull/2044

Reply:
Looks great, thanks!